ChopChop/Storage/FirebaseDatabase.swift

```diff
@@ -0,0 +1,242 @@
+            })
+            .eraseToAnyPublisher()
+    }
+
+    func updateRecipeRating(userId: String, recipeId: String, newScore: RatingScore) {
+        let docRef = db.collection(recipePath).document(recipeId)
+        db.runTransction({ transaction -> Any? in
+            let document: DocumentSnapshot
+            do {
+                try document = transaction.getDocument(docRef)
+            } catch {
+                return nil
+            }
+            guard let recipeRatings = try? document.data(as: OnlineRecipeRecord.self)?.ratings else {
+                return nil
+            }
+
+            var newRatingList = [RecipeRating]()
+
+            for rating in recipeRatings {
+                if rating.userId == userId {
+                    newRatingList.append(RecipeRating(userId: userId, score: newScore))
+                } else {
+                    newRatingList.append(rating)
+                }
+            }
+
+            transaction.updateData(["ratings": newRatingList.map({ $0.asDict })], forDocument: docRef)
+            return nil
+        })
+    }
+
+    func updateUserRating(userId: String, recipeId: String, newScore: RatingScore) {
+        let docRef = db.collection(userPath).document(userId)
+        db.runTransction({ transaction -> Any? in
+            let document: DocumentSnapshot
+            do {
+                try document = transaction.getDocument(docRef)
+            } catch {
+                return nil
+            }
+            guard let userRatings = try? document.data(as: User.self)?.ratings else {
+                return nil
+            }
+
+            var newRatingList = [UserRating]()
+
+            for rating in userRatings {
+                if rating.recipeId == recipeId {
+                    newRatingList.append(UserRating(recipeOnlineId: recipeId, score: newScore))
+                } else {
+                    newRatingList.append(rating)
+                }
+            }
+
+            transaction.updateData(["ratings": newRatingList.map({ $0.asDict })], forDocument: docRef)
+            return nil
+        })
+    }
+
+    func fetchOnlineRecipeById(onlineRecipeId: String) -> AnyPublisher<OnlineRecipeRecord, Error> {
+
+        db.collection(recipePath).document(onlineRecipeId)
+            .publisher()
+            .compactMap({ try? $0.data(as: OnlineRecipeRecord.self) })
+            .eraseToAnyPublisher()
+    }
+
+    func fetchOnlineRecipeOnceById(onlineRecipeId: String) -> AnyPublisher<OnlineRecipeRecord, Error> {
+        return db.collection(recipePath).document(onlineRecipeId)
+            .getDocument(as: OnlineRecipeRecord.self)
+            .compactMap {
+                $0
+            }
+            .eraseToAnyPublisher()
+    }
+
+    func addRecipeRating(onlineRecipeId: String, rating: RecipeRating) {
+        // swiftlint:disable implicit_return
+        return db.collection(recipePath).document(onlineRecipeId)
```

Comment:
Why do these functions return a value even though the function signature seems to indicate no return value

Reply:
Cos there are two updateData methods with different return signature, one that is a void and the other is a publisher. So I need to return in order for the compiler to know that I want to use the one that returns void cos my addRecipeRating is void. Taking out the return leads to compilation error due to ambiguous updateData that is used. Not sure if there is a better work around.